README.md
# [Nostr Citadel - The Sovereign Relay](https://github.com/MrJohnsson77/nostr-citadel)  

Nostr Citadel, a personal relay that stores and safeguards all your nostr data.  

The aim is to make running a relay and managing your data easy and accessible for everyone. 
Once you have set up your relay, it will automatically synchronize data from other relays and create a backup of your data locally, 
ensuring you never lose any events or notes.

By default, the relay will sync admin, paid, and whitelisted npubs. 
However, only admin, whitelisted, and paid npubs will have write access to the relay, unless you choose to set it as public in the configuration settings.

You can try it out by adding `wss://relay.nostr-citadel.org` to your relays.  
Stay Sovereign 🤙

![Nostr Citadel](screenshots/nostr-citadel-home-small.png?raw=true "Nostr Citadel Home")

![Startup Screen](screenshots/startup.png?raw=true "Startup Screen")

## Disclaimer
This is me learning [Nostr](https://github.com/fiatjaf/nostr) and [Go](https://go.dev/).  
I will be adding the features that I'd like to see in a sweet relay.

Let me know if there's anything you'd like to see implemented.

No prior Go experience, use it at your own peril. 💀  
It should be [safe](https://www.youtube.com/watch?v=dQw4w9WgXcQ) to use. 

There will be breaking changes.  
As Is - No Warranty!

## Credits
Thanks [fiatjaf](https://github.com/fiatjaf/relayer) for blueprint and inspiration. 💜

## Features / Todo

- [x] Nostr Relay
    * Core event model
- [x] Npub Whitelist
    * Whitelist your friends and foes for event posting
      * Limit reading in future version for private relays
- [x] Sync data to local relay
    * Sync either all whitelisted npubs or only admin data (notes) to the relay
    * The relay will sync every hour.
      * Using `since` to get new notes only. 
- [x] Cli
  * --help
  * --start
  * --port
  * --whitelist -h
  * --invoice -h
  * --backup -h
- [x] Simple Dashboard
  * Grid of active profiles on the relay
  * Can be disabled in config
- [x] Bootstrap Relays
- [x] Paid Relay
  * Core Lightning
  * LND
- [x] Simple Automated Backup
  - Automatic daily full backup
  - Simple CSV Format
  - Restore from CLI
- [ ] Improve Dashboard & Invoice Page
- [ ] Cleanup Workers
- [ ] Blacklist
  - IP / CIDR
- [ ] Automatic SSL Termination
  * Lets Encrypt Certificate
  * Self Issued Certificate
- [ ] Citadel Admin & Nostr Web Client (Separate Project - Coming "soon")
  * Web & Mobile (pwa) Client - [Nostr Citadel Web Client](https://github.com/MrJohnsson77/nostr-citadel-watch)
    * Nostr Client & Admin Dashboard for Nostr Citadel
    * Standard client functionality
      * Global
      * Followers
      * Groups
      * Notifications
      * Search
      * ... more ...
    * Follow recommendations
    * Trending on your relay ( and others... )
    * AI support, tweak and build your own recommendation engine
      * Share your own "models"
- [ ] Additional database backends

  
## Nips

[NIPs](https://github.com/nostr-protocol/nips) with a relay-specific implementation are listed here.

- [x] NIP-01: [Basic protocol flow description](https://github.com/nostr-protocol/nips/blob/master/01.md)
- [ ] NIP-02: [Contact List and Petnames](https://github.com/nostr-protocol/nips/blob/master/02.md)
- [ ] NIP-03: [OpenTimestamps Attestations for Events](https://github.com/nostr-protocol/nips/blob/master/03.md)
- [ ] NIP-05: [Mapping Nostr keys to DNS-based internet identifiers](https://github.com/nostr-protocol/nips/blob/master/05.md)
- [x] NIP-09: [Event Deletion](https://github.com/nostr-protocol/nips/blob/master/09.md)
- [x] NIP-11: [Relay Information Document](https://github.com/nostr-protocol/nips/blob/master/11.md)
- [x] NIP-11a: Relay Information Document Extensions
- [x] NIP-12: [Generic Tag Queries](https://github.com/nostr-protocol/nips/blob/master/12.md)
- [x] NIP-15: [End of Stored Events Notice](https://github.com/nostr-protocol/nips/blob/master/15.md)
- [x] NIP-16: [Event Treatment](https://github.com/nostr-protocol/nips/blob/master/16.md)
- [x] NIP-20: [Command Results](https://github.com/nostr-protocol/nips/blob/master/20.md)
- [ ] NIP-22: [Event `created_at` limits](https://github.com/nostr-protocol/nips/blob/master/22.md)
- [ ] NIP-26: [Event Delegation](https://github.com/nostr-protocol/nips/blob/master/26.md) 
- [ ] NIP-28: [Public Chat](https://github.com/nostr-protocol/nips/blob/master/28.md)
- [ ] NIP-33: [Parameterized Replaceable Events](https://github.com/nostr-protocol/nips/blob/master/33.md)
- [ ] NIP-42: [Authentication of clients to relays](https://github.com/nostr-protocol/nips/blob/master/42.md)

## Requirements
- Computer
- Internet
- Courage

## Get Started
Download the config.yaml and the binary for your architecture from the [release](https://github.com/MrJohnsson77/nostr-citadel/releases) section.  
Add your npub and relay_url in the config and drop the config and executable in a folder and start it.  

First start will create the database and bootstrap the relay by syncing the admin notes from other relays.
Initial list of relays is downloaded from [nostr.watch](https://api.nostr.watch/v1/online).

A specific bootstrap relay can be set in config.yaml, this relay will be added to the list of relays used during bootstrap.

Open the relay_url in a browser to verify that the relay is running.
Add your relay in your nostr client to connect and start saving your events.  
Profiles of whitelisted users will be displayed on the dashboard, it can be disabled in config.yaml by setting `dashboard: false`  

If your profile doesn't show up on the dashboard, start any nostr client and save your profile to push it to your relay.

### Operation
Changing admin.npub in config.yaml will replace the old admin with the new one.  
In this version a change of admin won't purge the events of the old admin, only delete the event with kind 0 (profile)

 *  Bind to port and select loglevel
    ```
    ./nostr-citadel start --port 1337 --loglevel INFO
    ```

  * Display whitelist
    ```
    # List whitelisted npubs  
    
    ./nostr-citadel whitelist --list
    ```

  * Add npub to whitelist
    ```
    ./nostr-citadel whitelist --add npub....
    ```

  * Remove npub from whitelist
    ```
    # Removing a npub from the whitelist will delete all events for it too.  
    
    ./nostr-citadel whitelist --remove npub....
    ```

  * Create invoice for npub
    ```
    ./nostr-citadel invoice --create npub....
    ```
    
  * Create QR Code invoice for npub
    ```
    ./nostr-citadel invoice --qr npub....
    ```
    
  * Verify if invoice is paid
    ```
    ./nostr-citadel invoice --verify npub....
    ```

  * Backup Relay
    ```
    ./nostr-citadel backup --all
    ```

  * Restore Backup to Relay
    ```
    ./nostr-citadel backup --restore backup/citadel-backup-events-20230407.csv
    ```
  
### Run from binary
Download binary for your architecture from the [releases](https://github.com/MrJohnsson77/nostr-citadel/releases) section.

### Run from source
  ```
  $ git clone git@github.com:MrJohnsson77/nostr-citadel.git
  $ cd nostr-citadel
  $ go run main
  ```

### Build from source
  ```
  $ git clone git@github.com:MrJohnsson77/nostr-citadel.git
  $ cd nostr-citadel
  $ make build
  ```

# Author

- [MrJohnsson](https://github.com/MrJohnsson77) - npub1fhdx6c6pt0ff6k3h5em760fzzzcehe9kqnjl05d2xwmg0ctjp80sn4hhsv 


# Contributors (A-Z)

- 

## License

This is free and unencumbered software released into the public domain.  
For more information, please refer to <http://unlicense.org/>

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.0.4] - 2023-04-08

### Added
- Backup and restore functionality
- Automated daily backups

## [v0.0.3] - 2023-04-03

### Added
- LND support for paid relay
- More logging
- More Config

## [v0.0.2] - 2023-03-20

### Added

- Config
  - Changed .yml to .yaml
  - Added more config settings
- Paid Relay
  - Core lightning supported
- Command line tool
  - whitelist
    - list
    - add
    - remove
  - invoice
    - create
    - verify
  - port
  - loglevel

### Fixed
- Project Structure

## [v0.0.1] - 2023-03-01

### Added

- Initial alpha version of nostr citadel.

[v0.0.1]: https://github.com/mrjohnsson77/nostr-citadel/releases/tag/v0.0.1
[v0.0.2]: https://github.com/mrjohnsson77/nostr-citadel/releases/tag/v0.0.2
[v0.0.3]: https://github.com/mrjohnsson77/nostr-citadel/releases/tag/v0.0.3
[v0.0.4]: https://github.com/mrjohnsson77/nostr-citadel/releases/tag/v0.0.4

handlers/nostr.go
package handlers

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"github.com/nbd-wtf/go-nostr"
	"github.com/nbd-wtf/go-nostr/nip19"
	"github.com/olahol/melody"
	"github.com/spf13/viper"
	"log"
	"nostr-citadel/storage"
)

type nip11Amount []struct {
	Amount int    `json:"amount"`
	Unit   string `json:"unit"`
}

type SpecialNIP11 struct {
	Contact     string `json:"contact"`
	Description string `json:"description"`
	Fees        struct {
		Admission   nip11Amount   `json:"admission"`
		Publication []interface{} `json:"publication"`
	} `json:"fees"`
	Limitation struct {
		AuthRequired     bool `json:"auth_required"`
		MaxContentLength int  `json:"max_content_length"`
		MaxEventTags     int  `json:"max_event_tags"`
		MaxFilters       int  `json:"max_filters"`
		MaxLimit         int  `json:"max_limit"`
		MaxMessageLength int  `json:"max_message_length"`
		MaxSubidLength   int  `json:"max_subid_length"`
		MaxSubscriptions int  `json:"max_subscriptions"`
		MinPowDifficulty int  `json:"min_pow_difficulty"`
		MinPrefix        int  `json:"min_prefix"`
		PaymentRequired  bool `json:"payment_required"`
	} `json:"limitation"`
	Name          string `json:"name"`
	PaymentsURL   string `json:"payments_url"`
	Pubkey        string `json:"pubkey"`
	Software      string `json:"software"`
	SupportedNips []int  `json:"supported_nips"`
	Version       string `json:"version"`
}

func GetPubKey(pubKey string) (string, string) {
	if len(pubKey) < 4 {
		return "", ""
	}
	var pub, npub string
	if pubKey[:4] == "npub" {
		_, v, err := nip19.Decode(pubKey)
		if err != nil {
			pub = ""
			npub = ""
		} else {
			pub = v.(string)
			npub = pubKey
		}
	} else {
		pub = pubKey
		pk, err := nip19.EncodePublicKey(pub)
		if err != nil {
			pub = ""
			npub = ""
		} else {
			npub = pk
		}
	}
	return pub, npub
}

func acceptEvent(event nostr.Event) bool {

	if viper.GetBool("relay_config.public_relay") {
		return true
	}
	wl := storage.GetWhitelisted(event.PubKey)
	whitelisted := event.PubKey == wl.PubKey
	nPub, _ := nip19.EncodePublicKey(event.PubKey)

	if !whitelisted {
		log.Println("Blocked event from:", nPub)
		return false
	}

	log.Println("Accepted event from:", nPub)
	return true
}

func NostrEventHandler(eventType string, requestData []json.RawMessage, s *melody.Session) error {

	switch eventType {
	case "EVENT":

		var event nostr.Event
		if err := json.Unmarshal(requestData[1], &event); err != nil {
			//notice = "failed to decode event: " + err.Error()
			return nil
		}

		serialized := event.Serialize()

		// Assign ID
		hash := sha256.Sum256(serialized)
		event.ID = hex.EncodeToString(hash[:])

		// check signature (requires the ID to be set)
		if ok, err := event.CheckSignature(); err != nil {
			log.Println("NOTICE", "error: failed to verify signature")
			sn, _ := json.Marshal([]interface{}{"OK", event.ID, false, "error: failed to verify signature"})
			_ = s.Write(sn)
			return nil
		} else if !ok {
			log.Println("NOTICE", "signature is invalid")
			sn, _ := json.Marshal([]interface{}{"OK", event.ID, false, "invalid: signature is invalid"})
			_ = s.Write(sn)
			return nil
		}

		// Check if event is accepted
		if !acceptEvent(event) {
			// NIP-20
			sn, _ := json.Marshal([]interface{}{"OK", event.ID, false, "blocked: no event posting access, contact admin."})
			_ = s.Write(sn)
			return nil
		}

		ok, message := storage.WriteEvent(&event)

		if ok {
			evr, _ := s.Get("events_sent")
			evRec := evr.(int)
			s.Set("events_sent", evRec+1)
			notifySubscribers(&event)
		}
		msg, _ := json.Marshal([]interface{}{"OK", event.ID, ok, message})
		_ = s.Write(msg)
		return nil
	case "REQ":
		var id string
		_ = json.Unmarshal(requestData[1], &id)
		if id == "" {
			log.Println("NOTICE", "invalid: Request has no ID")
			sn, _ := json.Marshal([]interface{}{"NOTICE", "invalid: Request has no ID"})
			_ = s.Write(sn)
			return nil
		}
		filters := make(nostr.Filters, len(requestData)-2)
		for i, filterReq := range requestData[2:] {
			if err := json.Unmarshal(
				filterReq,
				&filters[i],
			); err != nil {
				log.Println("NOTICE", "Failed to decode filter")
				sn, _ := json.Marshal([]interface{}{"NOTICE", "Failed to decode filter"})
				_ = s.Write(sn)
				return nil
			}

			filter := &filters[i]

			// Todo: Implement NIP-42 - Only allow authed users to get their private messages (kind-4)

			events, err := storage.GetEventsQuery(filter)
			if err != nil {
				log.Printf("Req Error: %v", err)
				continue
			}

			// Backup if query is broken.
			if filter.Limit > 0 && len(events) > filter.Limit {
				events = events[0:filter.Limit]
			}

			evr, _ := s.Get("events_received")
			evRec := evr.(int)
			s.Set("events_received", len(events)+evRec)

			for _, event := range events {
				sn, _ := json.Marshal([]interface{}{"EVENT", id, event})
				_ = s.Write(sn)
			}

		}
		setSubscriber(id, s, filters)
		sn, _ := json.Marshal([]interface{}{"EOSE", id})
		_ = s.Write(sn)
		return nil
	case "CLOSE":
		var id string
		_ = json.Unmarshal(requestData[1], &id)
		if id == "" {
			return nil
		}
		removeSubscriberId(s, id)
		return nil
	case "AUTH":
		// Todo: NIP-42
		return nil
	default:
		//reqData, _ := json.MarshalIndent(&requestData, "", "   ")
		//_ = SocketSend(reqData)
		return nil
	}
}

func NostrNip11() interface{} {

	supportedNIPs := []int{9, 11, 12, 15, 16, 20}

	//if authNip42 {
	//	supportedNIPs = append(supportedNIPs, 42)
	//}

	pubKey, _ := GetPubKey(viper.GetString("relay_config.admin_npub"))
	relayUrl := viper.GetString("relay_config.relay_url")

	if len(relayUrl) > 1 {
		relayUrl = relayUrl + "/invoices"
	}

	nip11Info := &SpecialNIP11{
		Name:          viper.GetString("relay_config.name"),
		Description:   viper.GetString("relay_config.description"),
		Pubkey:        pubKey,
		Contact:       viper.GetString("relay_config.admin_email"),
		SupportedNips: supportedNIPs,
		Software:      "git+https://github.com/MrJohnsson77/nostr-citadel.git",
		Version:       "0.0.1-alpha",
		PaymentsURL:   relayUrl,
	}

	// Todo: Connect it..
	nip11Info.Limitation.MaxMessageLength = 262144
	nip11Info.Limitation.MaxSubscriptions = 10
	nip11Info.Limitation.MaxFilters = 10
	nip11Info.Limitation.MaxLimit = 5000
	nip11Info.Limitation.MaxSubidLength = 500
	nip11Info.Limitation.MinPrefix = 4
	nip11Info.Limitation.MaxEventTags = 2500
	nip11Info.Limitation.MaxContentLength = 102400
	nip11Info.Limitation.MinPowDifficulty = 0
	nip11Info.Limitation.AuthRequired = false
	nip11Info.Limitation.PaymentRequired = true
	nip11Info.Fees.Admission = nip11Amount{
		{
			Amount: 50000,
			Unit:   "msats",
		},
	}
	nip11Info.Fees.Publication = []interface{}{}

	return nip11Info
}
